File: BlogApp-RN/src/features/authentication/data/domain/authRepoImplementation.ts
import { handleException } from "@/src/core/errors/handleException";
import { left, right } from "fp-ts/lib/Either";
import { AsyncResult } from "../../../../core/typedefs/asyncResult";
import AuthRepo from "../../domain/repos/authRepo";
import AuthDatasource from "../datasources/authDatasource";
import { AuthResponse } from "../models/authResponse";
import { ResponseMessage } from "../models/responseMessage";

class AuthRepoImplementation implements AuthRepo {
  constructor(private datasource: AuthDatasource) {}
  login: (username: string, password: string) => AsyncResult<AuthResponse> =
    async (username: string, password: string) => {
      try {
        const result = await this.datasource.login(username, password);
        return right(result);
      } catch (exception) {
        return left(handleException(exception));
      }
    };

  logout: () => AsyncResult<ResponseMessage> = async () => {
    try {
      const result = await this.datasource.logout();
      return right(result);
    } catch (exception) {
      return left(handleException(exception));
    }
  };

  signUp: (username: string, password: string) => AsyncResult<AuthResponse> =
    async (username: string, password: string) => {
      try {
        const result = await this.datasource.signUp(username, password);
        return right(result);
      } catch (exception) {
        return left(handleException(exception));
      }
    };
}

export default AuthRepoImplementation;

File: blogapp_flutter/lib/features/authentication/data/domain/auth_repo_implementation.dart
import 'package:blogapp_flutter/core/errors/handle_exception.dart';
import 'package:blogapp_flutter/core/typedefs/async_result.dart';
import 'package:blogapp_flutter/features/authentication/data/datasources/auth_datasource.dart';
import 'package:blogapp_flutter/features/authentication/domain/repos/auth_repo.dart';
import 'package:dartz/dartz.dart';

import '../models/auth_response.dart';
import '../models/response_message.dart';

class AuthRepoImplementation implements AuthRepo {
  final AuthDatasource datasource;
  late String url;

  AuthRepoImplementation(this.datasource);

  @override
  AsyncResult<AuthResponse> login({
    required String username,
    required String password,
  }) async {
    try {
      final response = await datasource.login(username, password);
      return Right(response);
    } catch (exception) {
      return Left(handleException(exception));
    }
  }

  @override
  AsyncResult<AuthResponse> signUp(
      {required String username, required String password}) async {
    try {
      final response = await datasource.signUp(username, password);
      return Right(response);
    } catch (exception) {
      return Left(handleException(exception));
    }
  }

  @override
  AsyncResult<ResponseMessage> logout() async {
    try {
      final response = await datasource.logout();
      return Right(response);
    } catch (exception) {
      return Left(handleException(exception));
    }
  }
}
